tritium-metrics/src/main/java/com/palantir/tritium/metrics/InstrumentedSslSocketFactory.java

```diff
@@ -0,0 +1,108 @@
+import java.net.Socket;
+import javax.net.ssl.HandshakeCompletedListener;
+import javax.net.ssl.SSLSocket;
+import javax.net.ssl.SSLSocketFactory;
+
+final class InstrumentedSslSocketFactory extends SSLSocketFactory {
+
+    private final SSLSocketFactory delegate;
+    private final HandshakeCompletedListener listener;
+
+    InstrumentedSslSocketFactory(SSLSocketFactory delegate, TaggedMetricRegistry metrics, String name) {
+        this.delegate = delegate;
+        this.listener = newHandshakeListener(metrics, name, "socket");
+    }
+
+    @Override
+    public String[] getDefaultCipherSuites() {
+        return delegate.getDefaultCipherSuites();
+    }
+
+    @Override
+    public String[] getSupportedCipherSuites() {
+        return delegate.getSupportedCipherSuites();
+    }
+
+    @Override
+    public Socket createSocket(Socket socket, String host, int port, boolean autoClose) throws IOException {
+        return wrap(delegate.createSocket(socket, host, port, autoClose));
+    }
+
+    @Override
+    public Socket createSocket(String host, int port) throws IOException {
+        return wrap(delegate.createSocket(host, port));
+    }
+
+    @Override
+    public Socket createSocket(String host, int port, InetAddress inetAddress, int clientPort)
+            throws IOException {
+        return wrap(delegate.createSocket(host, port, inetAddress, clientPort));
+    }
+
+    @Override
+    public Socket createSocket(InetAddress inetAddress, int port) throws IOException {
+        return wrap(delegate.createSocket(inetAddress, port));
+    }
+
+    @Override
+    public Socket createSocket(InetAddress inetAddress, int port, InetAddress clientAddress, int clientPort)
+            throws IOException {
+        return wrap(delegate.createSocket(inetAddress, port, clientAddress, clientPort));
+    }
+
+    @Override
+    public Socket createSocket(Socket socket, InputStream consumed, boolean autoClose) throws IOException {
+        return wrap(delegate.createSocket(socket, consumed, autoClose));
+    }
+
+    @Override
+    public Socket createSocket() throws IOException {
+        return wrap(delegate.createSocket());
+    }
+
+    @Override
+    public String toString() {
+        return "InstrumentedSSLSocketFactory{delegate=" + delegate + '}';
+    }
+
+    private Socket wrap(Socket socket) {
+        if (socket instanceof SSLSocket) {
+            ((SSLSocket) socket).addHandshakeCompletedListener(listener);
+        }
+        return socket;
+    }
+
+    static HandshakeCompletedListener newHandshakeListener(
+            TaggedMetricRegistry metrics, String name, String type) {
+        return event -> metrics.meter(MetricName.builder()
+                .safeName("tls.handshake")
+                .putSafeTags("name", name)
+                .putSafeTags("cipher", event.getCipherSuite())
+                .putSafeTags("protocol", event.getSession().getProtocol())
+                .build())
+                .mark();
```

Comment:
do you want to include a tag for `type`?

Reply:
Originally I included a type tag, but decided to remove it because I think the name tag will give us the information we want.

Removed in https://github.com/palantir/tritium/commit/a1b544721fa350072404c849149f036bf45d441e

Reply:
the `String type` method parameter is unused now, should we remove it?

Reply:
doh, fixed. Thanks for catching.